File: sample/src/main/java/cz/adaptech/tesseract4android/sample/Assets.java
package cz.adaptech.tesseract4android.sample;

import android.content.Context;
import android.content.res.AssetManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public class Assets {

    /**
     * Returns locally accessible directory where our assets are extracted.
     */
    @NonNull
    public static File getLocalDir(@NonNull Context context) {
        return context.getFilesDir();
    }

    /**
     * Returns locally accessible directory path which contains the "tessdata" subdirectory
     * with *.traineddata files.
     */
    @NonNull
    public static String getTessDataPath(@NonNull Context context) {
        return getLocalDir(context).getAbsolutePath();
    }

    @NonNull
    public static File getImageFile(@NonNull Context context) {
        return new File(getLocalDir(context), Config.IMAGE_NAME);
    }

    @Nullable
    public static Bitmap getImageBitmap(@NonNull Context context) {
        return BitmapFactory.decodeFile(getImageFile(context).getAbsolutePath());
    }

    public static void extractAssets(@NonNull Context context) {
        AssetManager am = context.getAssets();

        File localDir = getLocalDir(context);
        if (!localDir.exists() && !localDir.mkdir()) {
            throw new RuntimeException("Can't create directory " + localDir);
        }

        File tessDir = new File(getTessDataPath(context), "tessdata");
        if (!tessDir.exists() && !tessDir.mkdir()) {
            throw new RuntimeException("Can't create directory " + tessDir);
        }

        // Extract all assets to our local directory.
        // All *.traineddata into "tessdata" subdirectory, other files into root.
        try {
            for (String assetName : am.list("")) {
                final File targetFile;
                if (assetName.endsWith(".traineddata")) {
                    targetFile = new File(tessDir, assetName);
                } else {
                    targetFile = new File(localDir, assetName);
                }
                if (!targetFile.exists()) {
                    copyFile(am, assetName, targetFile);
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static void copyFile(@NonNull AssetManager am, @NonNull String assetName,
                                 @NonNull File outFile) {
        try (
                InputStream in = am.open(assetName);
                OutputStream out = new FileOutputStream(outFile)
        ) {
            byte[] buffer = new byte[1024];
            int read;
            while ((read = in.read(buffer)) != -1) {
                out.write(buffer, 0, read);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: sample/src/main/java/cz/adaptech/tesseract4android/sample/Assets.kt
package cz.adaptech.tesseract4android.sample

import android.content.Context
import android.content.res.AssetManager
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

object Assets {
	/**
	 * Returns locally accessible directory where our assets are extracted.
	 */
	fun getLocalDir(context: Context): File {
		return context.filesDir
	}

	/**
	 * Returns locally accessible directory path which contains the "tessdata" subdirectory
	 * with *.traineddata files.
	 */
    @JvmStatic
    fun getTessDataPath(context: Context): String {
		return getLocalDir(context).absolutePath
	}

	@JvmStatic
    fun getImageFile(context: Context): File {
		return File(getLocalDir(context), Config.IMAGE_NAME)
	}

	@JvmStatic
    fun getImageBitmap(context: Context): Bitmap? {
		return BitmapFactory.decodeFile(getImageFile(context).absolutePath)
	}

	@JvmStatic
    fun extractAssets(context: Context) {
		val am = context.assets

		val localDir = getLocalDir(context)
		if (!localDir.exists() && !localDir.mkdir()) {
			throw RuntimeException("Can't create directory $localDir")
		}

		val tessDir = File(getTessDataPath(context), "tessdata")
		if (!tessDir.exists() && !tessDir.mkdir()) {
			throw RuntimeException("Can't create directory $tessDir")
		}

		// Extract all assets to our local directory.
		// All *.traineddata into "tessdata" subdirectory, other files into root.
		try {
			for (assetName in am.list("")!!) {
				val targetFile = if (assetName.endsWith(".traineddata")) {
					File(tessDir, assetName)
				} else {
					File(localDir, assetName)
				}
				if (!targetFile.exists()) {
					copyFile(am, assetName, targetFile)
				}
			}
		} catch (e: IOException) {
			e.printStackTrace()
		}
	}

	private fun copyFile(
		am: AssetManager, assetName: String,
		outFile: File
	) {
		try {
			am.open(assetName).use { `in` ->
				FileOutputStream(outFile).use { out ->
					val buffer = ByteArray(1024)
					var read: Int
					while ((`in`.read(buffer).also { read = it }) != -1) {
						out.write(buffer, 0, read)
					}
				}
			}
		} catch (e: IOException) {
			e.printStackTrace()
		}
	}
}
